### PR TITLE
glib-macros: Pass a pointer for watched objects into closure macro

### PR DESCRIPTION
**Pull request by jf2048**
This also saves us from doing an extra `ref`/`unref` on watched objects when calling those closures

### CODE REVIEW

File: glib/src/object.rs [hunk 1]
```diff
@@ -3168,6 +3168,50 @@ impl<T: ObjectType> ObjectExt for T {
     }
 }
 
+// Helper struct to avoid creating an extra ref on objects inside closure watches. This is safe
+// because `watch_closure` ensures the object has a ref when the closure is called.
+#[doc(hidden)]
+pub struct WatchedObject<T: ObjectType>(*mut T::GlibType);
```
Comment (sdroege) [inline glib/src/object.rs:3174]:
```suggestion
pub struct WatchedObject<T: ObjectType>(*mut T::GlibType);

unsafe impl<T: ObjectType + Send + Sync> Send for WatchedObject {}
```

File: glib/src/object.rs [hunk 2]
```diff
@@ -3168,6 +3168,50 @@ impl<T: ObjectType> ObjectExt for T {
...
+    where
+        T: FromGlibPtrBorrow<*mut <T as ObjectType>::GlibType>,
+    {
+        from_glib_borrow(self.0)
+    }
+}
+
+#[doc(hidden)]
+pub trait Watchable<T: ObjectType> {
+    fn get_object(&self) -> WatchedObject<T>;
```
Comment (sdroege) [inline glib/src/object.rs:3191]:
```suggestion
    fn get_watched_object(&self) -> WatchedObject<T>;
```

Maybe?

File: glib-macros/tests/test.rs [hunk 3]
```diff
@@ -359,7 +359,7 @@ fn closure() {
         assert_eq!(obj.ref_count(), 1);
         let weak_test = glib::closure_local!(@watch obj => move || obj.ref_count());
         assert_eq!(obj.ref_count(), 1);
-        assert_eq!(weak_test.invoke::<u32>(&[]), 3);
+        assert_eq!(weak_test.invoke::<u32>(&[]), 2);
```
Comment (sdroege) [inline glib-macros/tests/test.rs:362]: Why? :)

**Review by sdroege — COMMENTED** [review]
I was trying to get around that without a new trait, but I couldn't find a way. You apparently also couldn't :) Oh well! At least it doesn't have to be in the prelude

File: glib/src/object.rs (hunk 2 above)
Comment (bilelmoussaoui) [reply glib/src/object.rs:3191]: maybe even `watched_object`

File: glib/src/object.rs (hunk 1 above)
Comment (sdroege) [reply glib/src/object.rs:3174]: Should probably also use `std::ptr::NonNull` here

**Review by sdroege — APPROVED** [review]

**Pull request merged**